[enh] improve openstreetmap engine

dalf opened the pull request:
## What does this PR do?

implements ideas described in https://github.com/searxng/searxng/issues/69

## Why is this change important?

* the master branch allow to make direct HTTP request from the browser to overpass API : it breaks the privacy idea.
* include wikidata data: an image, website & wikipedia links.
* the overpass data are now fetched directly using the nominatim API.

This PR is a draft / a demo  : 
* [x] there are some `style=...` in `map.html`
* [x] the simple theme is not updated. 
* [x] localization is not implemented.
* [x] the HTML layout can be improved.
* [x] the javascript to call the overpass API is not removed.
* [x] `searx/data/osm_tags.json` should be named  `searx/data/osm.json` and can be smaller.

## How to test this PR locally?

See https://wiki.openstreetmap.org/wiki/Nominatim/Special_Phrases

Some example:
* `!osm museum in berlin`
* `!osm ATMs in bruxelles`
* `!osm paris`
* `!osm cinémas in lyon`

## Author's checklist

<!-- additional notes for reviewiers -->

## Related issues

<!--
Closes #234
-->

dalf commented on searx/engines/wikidata.py:
```diff
@@ -64,6 +65,7 @@
         mwapi:language "%LANGUAGE%".
         ?item wikibase:apiOutputItem mwapi:item.
   }
+  MINUS { ?item wdt:P31 wd:Q21281405 . }
```
Not related to this PR....
(see line 55)

return42 commented on searx/engines/openstreetmap.py:
```diff
@@ -23,24 +29,176 @@
     return params
 
 
+def fetch_wikidata(nominatim_json, user_langage):
```
I think functions like this needs to be documented .. for a newbee it will take hours and hours to understand whats going on here .. most often I gave up after 15min. 

The criticism is not related to this PR it is more general, I often stuck because no function in searx is documented. I don't think every little thing needs to be documented, but things that are not obvious from the implementation should be documented. 

What I mean: in the future we should more document .. I am also addressing myself :)

dalf replied:
* Lines 131 to 138: 
  * set `wikidata_id` to the wikidata ids of `['extratags']['wikidata']` or `['extratags']['wikidata link']` (I have seen both...)
  * set `wd_to_results` to mapping from a wikidata id to a list of result (the actual result in `nominatim_json`).
* Lines 141 to 143: send the query using `wikidata_id`.
* Line 144: for each result of the SPARQL request (more or less for each wikidata id (*) ).
* line 146: update the results of `nominatim_json` using `wd_to_results` = for each result of `nominatim_json` which reference `wikidata_id`, set the `wikidata` entry.

(*) the result may contain more than one row per wikidata id (same issue than `SELECT * FROM a INNER JOIN b ON b.parent_id = a.id`, if there are two images, there are two rows for the same item). It is silently ignored in the code.

return42 commented on searx/engines/openstreetmap.py:
```diff
@@ -23,24 +29,176 @@
+                    or result.get('itemLabel', {}).get('value')
+
+
+def set_value(d, k, o, *args):
```
If I am not wrong, this function is not used.

dalf committed:
[fix] strip spaces from searx user agent

h11 (used by httpx) rejects HTTP request with a trailing space in HTTP headers

return42 commented on searx/static/themes/oscar/js/searx.min.js:
```diff
@@ -1,4 +1,4 @@
-/*! oscar/searx.min.js | 01-06-2021 | https://github.com/searxng/searxng  */
+/*! oscar/searx.min.js | 02-06-2021 | https://github.com/searxng/searxng  */
```
bikeshedding: this patch can be moved to the commit which includes the build products [static].

dalf replied:
Done.

dalf committed:
[enh] openstreetmap / map template: improve results

implements ideas described in #69

* update the engine
* use wikidata
* update map.html template

return42 commented on searx/data/__init__.py:
```diff
@@ -3,7 +3,7 @@
 
 __init__ = ['ENGINES_LANGUGAGES', 'CURRENCIES', 'USER_AGENTS', 'EXTERNAL_URLS', 'WIKIDATA_UNITS', 'EXTERNAL_BANGS',
-            'bangs_loader', 'ahmia_blacklist_loader']
+            'OSM_KEYS_TAGS', 'bangs_loader', 'ahmia_blacklist_loader']
```
@dalf do you know where this `__init__` variable is used?  I have problems to find the usage.

dalf replied:
It should `__all__ ` not `__init__` !

dalf replied:
updated

dalf committed:
[fix] searx/data/__init__.py: rename __init__ as __all__

return42 committed:
[pylint] searx/data/__init__.py

BTW: add doc strings and moved __all__ to the top [1]

[1] https://www.python.org/dev/peps/pep-0008/#module-level-dunder-names

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[pylint] searx_extra/update/update_osm_keys_tags.py

BTW: move some comments into script's  doc-string

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[data.all] add searx_extra/update/update_osm_keys_tags.py

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] openstreetmap - fix some minor whitespace & indentation issues

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] make simple/result_templates/map.html more CSP compliant [1]

[1] https://github.com/searxng/searxng/issues/57

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[static] make themes.all - based on commit 12af3b39

Based on commit 12af3b39 this patch is generated by::

    make themes.all

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>
